### PR TITLE
Support for 'CIS Docker Benchmark 1.12.0'

### PR DESCRIPTION
**Pull request by vadapalliravikumar**
Signed-off-by: Ravi Kumar Vadapalli <vadapalli.ravikumar@gmail.com>

The code changes are to support 'CIS Docker 1.12.0 Benchmark'. 17 new recommendations have been added in 1.12.0 benchmark. 12 out of these 17 have tests have been included in these code changes.

Tested in Ubuntu16.04 with GNU bash 4.3.46. Checked the changes with shellcheck & checkbashims.

### CODE REVIEW

**Comment by konstruktoid** [comment]
Thanks a lot @vadapalliravikumar! I'll go through it all in detail and update when CIS 1.13 is available, which should be Soon™

File: tests/5_container_runtime.sh [hunk 1]
```diff
@@ -561,8 +561,123 @@ else
...
+        warn "     * pidslimit not set: $c"
+      fi
+    fi
+  done
+  # We went through all the containers and found all with PIDs limit
+  if [ $fail -eq 0 ]; then
+      pass "$check_5_28"
+  fi
+
+  # 5.29
```
Comment (konstruktoid) [inline tests/5_container_runtime.sh]: This results in 
```
[WARN] 5.29 - Do not use Docker's default bridge docker0
[WARN]      * container in docker0 network:  16794ef13b98785b74c08775d3580b44cb7648f140a3cd3871e7d05f9512d852  219607ecdfe3a37161fcba50ae92c54dd1d6c12f465a434a9cec48e58e9eb61e
``` 
were the IDs and not names are presented on a single line.

File: tests/5_container_runtime.sh [hunk 2]
```diff
@@ -561,8 +561,123 @@ else
...
+    docker inspect --format '{{ .Mounts }}' "$c" 2>/dev/null | grep 'docker.sock' >/dev/null 2>&1
+
+    if [ $? -eq 0 ]; then
+      # If it's the first container, fail the test
+      if [ $fail -eq 0 ]; then
+        warn "$check_5_31"
+        warn "     * Docker sock shared: $c"
+        fail=1
+      else
+        warn "     * Docekr sock shared: $c"
```
Comment (konstruktoid) [inline tests/5_container_runtime.sh]: Typo.

File: tests/2_docker_daemon_configuration.sh [hunk 3]
```diff
@@ -134,3 +134,48 @@ if [ $? -eq 0 ]; then
...
+docker info 2>/dev/null | grep -e "Swarm:\s*active\s*" >/dev/null 2>&1
+if [ $? -eq 1 ]; then
+  pass "$check_2_15"
+else
+  warn "$check_2_15"
+fi
+
+# 2.16
+check_2_16="2.16 - Control the number of manager nodes in a swarm"
+docker node ls 2>/dev/null | grep "Leader" >/dev/null 2>&1
```
Comment (konstruktoid) [inline tests/2_docker_daemon_configuration.sh]: This doesn't count the actual number of leaders.

File: tests/4_container_images.sh [hunk 4]
```diff
@@ -38,10 +38,66 @@ fi
...
+fail=0
+for img in $images; do
+  docker history $img 2>/dev/null | grep -e "update" >/dev/null 2>&1
+  if [ $? -eq 0 ]; then
+    if [ $fail -eq 0 ]; then
+      fail=1
+      info "$check_4_7"
+    fi
+    imgName=`docker inspect --format='{{.RepoTags}}' $img 2>/dev/null`
+    info "      update instruction found in history of $imgName"
```
Comment (konstruktoid) [inline tests/4_container_images.sh]: Should start with capital letter.

File: tests/4_container_images.sh [hunk 5]
```diff
@@ -38,10 +38,66 @@ fi
...
+fail=0
+for img in $images; do
+  docker history $img 2> /dev/null | grep 'ADD' >/dev/null 2>&1
+  if [ $? -eq 0 ]; then
+    if [ $fail -eq 0 ]; then
+      fail=1
+      info "$check_4_9"
+    fi
+    imgName=`docker inspect --format='{{.RepoTags}}' $img 2>/dev/null`
+    info "      found ADD in docker history of $imgName"
```
Comment (konstruktoid) [inline tests/4_container_images.sh]: Should start with capital letter.

File: tests/5_container_runtime.sh [hunk 6]
```diff
@@ -561,8 +561,123 @@ else
...
+
+  fail=0
+  for c in $containers; do
+    pidslimit=`docker inspect --format '{{.HostConfig.PidsLimit }}' "$c"`
+
+    if [ $pidslimit -le 0 ]; then
+      # If it's the first container, fail the test
+      if [ $fail -eq 0 ]; then
+        warn "$check_5_28"
+        warn "     * pidslimit not set: $c"
```
Comment (konstruktoid) [inline tests/5_container_runtime.sh]: Should start with capital letter.

File: tests/5_container_runtime.sh [hunk 7]
```diff
@@ -561,8 +561,123 @@ else
...
+        fail=1
+      else
+        warn "     * pidslimit not set: $c"
```
Comment (konstruktoid) [inline tests/5_container_runtime.sh]: Should start with capital letter.

File: tests/5_container_runtime.sh [hunk 8]
```diff
@@ -561,8 +561,123 @@ else
...
+
+    if [ $? -eq 0 ]; then
+      docker0Containers=`docker network inspect --format='{{ range $k, $v := .Containers }} {{ $k }} {{ end }}' "$net" 2>/dev/null`
+      if [ -n "$docker0Containers" ]; then
+        if [ $fail -eq 0 ]; then
+          warn "$check_5_29"
+          fail=1
+        fi
+        for c in $docker0Containers; do
+          warn "     * container in docker0 network: $c"
```
Comment (konstruktoid) [inline tests/5_container_runtime.sh]: Should start with capital letter.

File: tests/2_docker_daemon_configuration.sh [hunk 9]
```diff
@@ -134,3 +134,48 @@ if [ $? -eq 0 ]; then
...
+if [ $? -eq 1 ]; then
+  pass "$check_2_16"
+else
+  warn "$check_2_16"
+fi
+
+# 2.17
+check_2_17="2.17 - Bind swarm services to a specific host interface"
```
Comment (konstruktoid) [inline tests/2_docker_daemon_configuration.sh]: This does not check if a specific host interface is used.
```
$ netstat -tl | grep 2377
tcp        0      0 10.2.3.41:2377          *:*                     LISTEN 
```

File: tests/4_container_images.sh [hunk 10]
```diff
@@ -38,10 +38,66 @@ fi
...
+fail=0
+for img in $images; do
+  docker inspect --format='{{.Config.Healthcheck}}' $img 2>/dev/null | grep -e "<nil>" >/dev/null 2>&1
+  if [ $? -eq 0 ]; then
+    if [ $fail -eq 0 ]; then
+      fail=1
+      warn "$check_4_6"
+    fi
+    imgName=`docker inspect --format='{{.RepoTags}}' $img 2>/dev/null`
+    warn "      No Healthcheck found : $imgName"
```
Comment (konstruktoid) [inline tests/4_container_images.sh]: Use consistent warning and info messages. Start with a * and capital letter.

**Comment by konstruktoid** [comment]
@vadapalliravikumar please see the comments, could you update your PR?

**Comment by konstruktoid** [comment]
Merging and we'll go from there, so you can get credit for the work you've done.

**Comment by vadapalliravikumar** [comment]
Thank you for merging the changes. Sorry about not being able to manage time to address the comments before the merge.

**Comment by konstruktoid** [comment]
You're welcome @vadapalliravikumar, take care.

**Pull request merged**